python/ray/tune/ray_trial_executor.py

```diff
@@ -188,6 +202,67 @@ def __init__(self,
         if ray.is_initialized():
             self._update_avail_resources()
 
+    def in_staging_grace_period(self) -> bool:
+        """Returns True if trials have recently been staged."""
+        return self._pg_manager.in_staging_grace_period()
+
+    def stage_and_update_status(self, trials: List[Trial]):
+        """Check and update statuses of scheduled placement groups.
+
+        Stages placement groups of all trials.
+        """
+        for trial in trials:
+            if trial.status != Trial.PENDING:
+                continue
+            if not trial.uses_placement_groups:
+                continue
+            if self._pg_manager.is_ready(trial):
+                continue
+            if self._pg_manager.is_staging(trial):
+                continue
+
+            if not self._pg_manager.stage_trial(trial):
+                # Break if we reached the limit of pending placement groups.
+                break
+
+        self._pg_manager.update_status()
+
+    def get_staged_trial(self, replace: Optional[Trial] = None):
```

Comment:
hmmm, is it ok not to implement this switching? 

I'm thinking maybe this logic isn't too consequential (i.e., it's not terrible if trials are started out of order)

Reply:
I implemented this to be sure that the FIFOScheduler actually still does first in first out. What is your main concern here if we leave it? Code complexity?

Reply:
Yeah I think it just seems like a bit complicated of a solution / could be hard for others to maintain in the future.

Reply:
I guess in terms of the "FIFO invariant", i'm pretty lax about that because in the case where you have 100 concurrent trials, it's expected that some trials will start out of order

Reply:
(mainly seeing the `switch_trial_pgs` implementation)

Reply:
I refactored the manager - see main comment below